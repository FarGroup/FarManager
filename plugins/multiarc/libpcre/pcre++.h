#pragma once

#include <string.h>
#include "pcre.h"

namespace PCRE
{

  class Match
  {
    int ovector[30];
    int matchCount;
    const pcre *re;
    const char *str;
    int strLen;
    int bufLen;
    char *buffer;

    friend class RegExp;

    Match(const pcre *re, const char *str)
      : re(re),
        str(str),
        strLen(0),
        bufLen(0),
        buffer(0)
    {
      for(const char *p = str; *p++; ++strLen);

      matchCount = pcre_exec(re, 0, str, strLen, 0, 0, ovector, sizeof(ovector) / sizeof(ovector[0]));

      if(matchCount > 0)
        buffer = (char*)pcre_malloc(bufLen = strLen + 10);
    }

  public:

    Match(const Match &m)
      : matchCount(m.matchCount),
        re(m.re),
        str(m.str),
        strLen(m.strLen),
        bufLen(m.bufLen),
        buffer(0)
    {
      for(size_t i = 0; i < sizeof(ovector) / sizeof(ovector[0]); ++i)
        ovector[i] = m.ovector[i];
      if(bufLen > 0)
        buffer = (char*)pcre_malloc(bufLen);
    }

    Match &operator=(const Match &m)
    {
      re = m.re;
      str = m.str;
      strLen = m.strLen;
      matchCount = m.matchCount;
      bufLen = m.bufLen;
      if(buffer)
        pcre_free(buffer);
      for(size_t i = 0; i < sizeof(ovector) / sizeof(ovector[0]); ++i)
        ovector[i] = m.ovector[i];
      if(bufLen > 0)
        buffer = (char*)pcre_malloc(bufLen);
    }

    operator bool()
    {
      return matchCount >= 0;
    }

    const char *operator[] (const char *name)
    {
      int res = pcre_copy_named_substring(re, str, ovector, matchCount, name, buffer, bufLen);
      return res >= 0 ? buffer : 0;
    }

    const char *operator[] (int idx)
    {
      int res = pcre_copy_substring(str, ovector, matchCount, idx, buffer, bufLen);
      return res >= 0 ? buffer : 0;
    }

    ~Match()
    {
      if(buffer)
        pcre_free(buffer);
    }
  };

  class RegExp
  {
    pcre *re;
    const char *lastError;
    int errorOffset;

    RegExp(const RegExp &);         // I'm lazy =)
    RegExp &operator=(const RegExp &);

  public:
    RegExp()
      : re(0),
        lastError(0),
        errorOffset(0)
    {
    }

    RegExp(const char *pattern)
      : re(0),
        lastError(0),
        errorOffset(0)
    {
      compile(pattern);
    }

    RegExp(const char *pattern, const char *options)
      : re(0),
        lastError(0),
        errorOffset(0)
    {
      compile(pattern, options);
    }

    ~RegExp()
    {
      if(re)
        pcre_free(re);
    }

    bool compile(const char *pattern)
    {
      RegExp r("^\\/(.*)\\/([ixms]*)$", "");

      if(Match m = r.match(pattern))
      {
        char options[8] = "";
        strncat(options, m[2], sizeof(options));
        return compile(m[1], options);
      }
      return false;
    }

    bool compile(const char *pattern, const char *options)
    {
      int nOptions = 0;
      for(const char *opt = options; opt && *opt; ++opt)
      {
        switch(*opt)
        {
          case 'i': nOptions |= PCRE_CASELESS;  break;
          case 'x': nOptions |= PCRE_EXTENDED;  break;
          case 'm': nOptions |= PCRE_MULTILINE; break;
          case 's': nOptions |= PCRE_DOTALL;  break;
        }
      }

      //const char *error = 0;
      //int errorOfs = 0;
      if(re)
        pcre_free(re);
      re = pcre_compile(pattern, nOptions, &lastError, &errorOffset, 0);

      return !!re;
    }

    bool isValid() const
    {
      return !!re;
    }

    const char *getLastError() const
    {
      return lastError;
    }

    int getErrorOffset() const
    {
      return errorOffset;
    }

    Match match(const char *str) const
    {
      return Match(re, str);
    }

  };


}
