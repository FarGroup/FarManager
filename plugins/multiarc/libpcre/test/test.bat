@echo off
copy ..\pcre.h .
copy "..\pcre++.h" .
copy ..\pcre.lib .
cl /nologo /GX /MD test.cpp

test.exe < test.in > test.out2

fc /b test.out test.out2
