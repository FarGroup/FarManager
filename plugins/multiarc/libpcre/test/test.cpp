#define PCRE_STATIC
#include "pcre++.h"
#include <iostream>
#include <string>
using namespace std;
using namespace PCRE;

#pragma comment(lib, "pcre.lib")

int main(int argc, char **argv)
{
	RegExp re(
		"^(?P<date>\\d\\d\\.\\d\\d\\.(\\d\\d)?\\d\\d)\\s+(?P<time>\\d\\d:\\d\\d)\\s+(?P<size>(\\d+\\xff?)+)\\s+(?P<name>[^\\s].*[^\\s])\\s*$",
		"ix");

	string input;
	while(getline(cin, input))
	{
		if(Match m = re.match(input.c_str()))
		{
			if(m["name"])
				cout << "---------------------" << endl << "name: " << m["name"] << endl;
			if(m["date"])
				cout << "date: " << m["date"] << endl;
			if(m["time"])
				cout << "time: " << m["time"] << endl;
			if(m["size"])
				cout << "size: " << m["size"] << endl;
		}
	}
	
	return 0;
}
