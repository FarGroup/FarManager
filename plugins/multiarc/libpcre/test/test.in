 Volume in drive D is work
 Volume Serial Number is xxxx-xxxx

 Directory of D:\works\pcre-4.4-cut\test

27.11.2004  01:28    <DIR>          .
27.11.2004  01:28    <DIR>          ..
27.11.2004  01:27             3ÿ507 pcre++.h
25.08.2003  22:30             6ÿ884 pcre.h
27.11.2004  01:24            96ÿ364 pcre.lib
27.11.2004  01:24               775 test.cpp
27.11.2004  01:27            45ÿ056 test.exe
27.11.2004  01:28                 0 test.in
27.11.2004  01:27            52ÿ231 test.obj
               7 File(s)        204ÿ817 bytes
               2 Dir(s)  37ÿ907ÿ562ÿ496 bytes free
