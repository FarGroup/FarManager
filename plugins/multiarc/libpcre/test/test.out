---------------------
name: pcre++.h
date: 27.11.2004
time: 01:27
size: 3ÿ507
---------------------
name: pcre.h
date: 25.08.2003
time: 22:30
size: 6ÿ884
---------------------
name: pcre.lib
date: 27.11.2004
time: 01:24
size: 96ÿ364
---------------------
name: test.cpp
date: 27.11.2004
time: 01:24
size: 775
---------------------
name: test.exe
date: 27.11.2004
time: 01:27
size: 45ÿ056
---------------------
name: test.in
date: 27.11.2004
time: 01:28
size: 0
---------------------
name: test.obj
date: 27.11.2004
time: 01:27
size: 52ÿ231
